src/rust/crates/ain-evm/src/handler.rs

```diff
@@ -13,4 +20,85 @@ impl Handlers {
             block: BlockHandler::new(),
         }
     }
+
+    pub fn finalize_block(
+        &self,
+        context: u64,
+        update_state: bool,
+    ) -> Result<(Block<TransactionV2>, Vec<TransactionV2>), Box<dyn Error>> {
+        let mut tx_hashes = Vec::with_capacity(self.evm.tx_queues.len(context));
+        let mut failed_tx_hashes = Vec::with_capacity(self.evm.tx_queues.len(context));
+        let vicinity = get_vicinity(None, None);
+        let state = self.evm.tx_queues.state(context).expect("Wrong context");
+        let backend = MemoryBackend::new(&vicinity, state);
+        let mut executor = AinExecutor::new(backend);
+
+        for signed_tx in self.evm.tx_queues.drain_all(context) {
+            let tx_response = executor.exec(&signed_tx);
+            if tx_response.exit_reason.is_succeed() {
+                tx_hashes.push(signed_tx.transaction);
+            } else {
+                failed_tx_hashes.push(signed_tx.transaction)
+            }
+        }
+
+        self.evm.tx_queues.remove(context);
+
+        if update_state {
+            let mut state = self.evm.state.write().unwrap();
+            *state = executor.backend().state().clone();
+        }
+
+        let mut parent_hash: H256 = Default::default();
+        let mut number: usize = 0;
+
+        let blocks = self.block.blocks.read().unwrap();
+        if !blocks.is_empty() {
+            parent_hash = blocks.first().unwrap().header.hash();
+            number = blocks.len() + 1;
+        }
+        drop(blocks);
```

Comment:
```suggestion
        let (parent_hash, number) = {
            let blocks = self.block.blocks.read().unwrap();
            blocks
                .first()
                .and_then(|first_block| Some((first_block.header.hash(), blocks.len() + 1)))
                .unwrap_or((H256::default(), 0))
        };
```
What about something like this so we remove the need to drop explicititly and the mut variables

Reply:
This looks excellent, will do this instead. Was unaware of this syntax in Rust.

Reply:
Fixed: https://github.com/DeFiCh/ain/pull/1882/commits/922dca0ccda9a4d6a384bda1f0ccc07fea88d777